### PR TITLE
resolve next config being deleted problem

### PR DESCRIPTION
**Pull request by ymchun**
Fixes https://github.com/shadowwalker/next-pwa/issues/371

changed
```js
const withPWA = require('next-pwa')

module.exports = withPWA({
  pwa: {
    dest: 'public'
  }
})
```

to
```js
const withPWA = require('next-pwa')
const nextConfig = {}

module.exports = withPWA(nextConfig, {
  dest: 'public'
})
```

### CODE REVIEW

**Comment by Lukasdotcom** [comment]
When I try this on my end(I manually copied the index.js file into my node_modules folder) it still enables next-pwa when I set the disable flag to true.
My server logs show this
```
ready - started server on 0.0.0.0:3000, url: http://localhost:3000
info  - Loaded env from /Users/lukas/Documents/Coding/bundesliga/.env.local
> [PWA] Compile client (static)
> [PWA] Auto register service worker with: /Users/lukas/Documents/Coding/bundesliga/node_modules/next-pwa/register.js
> [PWA] Service worker: /Users/lukas/Documents/Coding/bundesliga/.next/sw.js
> [PWA]   url: /sw.js
> [PWA]   scope: /
> [PWA] Build in develop mode, cache and precache are mostly disabled. This means offline support is disabled, but you can continue developing other functions in service worker.
> [PWA] Compile server
> [PWA] PWA support is disabled
> [PWA] Compile server
> [PWA] PWA support is disabled
```
Also when I visit the website with chrome this error is shown in console
```
Uncaught (in promise) DOMException: Failed to register a ServiceWorker for scope ('http://localhost:3000/') with script ('http://localhost:3000/sw.js'): Operation has been aborted
```
My next.config.js looks like this
```js
const withPWA = require("next-pwa");

/** @type {import('next').NextConfig} */
const nextConfig = withPWA({
  reactStrictMode: true
});

module.exports = withPWA(nextConfig, {
  dest: "public",
  disable: process.env.NODE_ENV !== "production",
})
```

**Comment by ymchun** [comment]
> 

```
/** @type {import('next').NextConfig} */
const nextConfig = withPWA({
  reactStrictMode: true
});
```
seems this part doesn't need to be wrapped with `withPWA`

**Review by Lukasdotcom — APPROVED** [review]
LGTM

**Comment by marcofranssen** [comment]
The `withPWA` function does only have one input parameter. How does this fix that?

So where does your second argument be handled?

```javascript
module.exports = (nextConfig = {})
```

**Comment by ymchun** [comment]
> 

therefore, I introduced a breaking change to take in the second argument

https://github.com/shadowwalker/next-pwa/blob/0ad2ea7c90563427085bb931f2181ce9a201a02c/index.js#L15

**Comment by marcofranssen** [comment]
Ah, I missed that part. However do we really need that breaking change? I think I'm my PR I fixed it without a breaking change unless I missed something.

**Comment by ymchun** [comment]
@marcofranssen whether having a breaking change at this moment is the choice of the maintainers. I would like to point out that both of our MRs are fixing the issue but which one is better solution and will last for longer time.

as the comment here pointed out Next.js will validate its config and warn for unknown properties. I would believe the reason behind this is to avoid putting third-party config into Next.js config.
https://github.com/shadowwalker/next-pwa/blob/9d28c3be6aa6cc082a2ca87ba61117fff57ded8f/index.js#L337

for your MR, you solve the config issue. My MR solve not only the config issue but also align to Next.js config convention.

P.S. Next.js config function from Sentry for your reference.
https://github.com/getsentry/sentry-javascript/blob/fbb44d5e1860840e65f36cf4811c31291d51e186/packages/nextjs/src/config/index.ts

**Comment by AleksandarDev** [comment]
If we want to introduce breaking change - can we change it to be like other plugins for nextjs? eg.

```js
const withMDX = require('@next/mdx')({
  options: {
    remarkPlugins: [],
    rehypePlugins: [],
  },
})
module.exports = withMDX(nextConfig)
```

or 

```js
const withBundleAnalyzer = require('@next/bundle-analyzer')({
    enabled: process.env.ANALYZE === 'true',
});
```

This way configuration for each plugin is on its own. 

With this PR we would get bad example version...

```js
// BAD EXAMPLE

const withBundleAnalyzer = require('@next/bundle-analyzer')({
    enabled: process.env.ANALYZE === 'true',
});

const withPWA = require('next-pwa');

module.exports = withBundleAnalyzer(withPWA(withMDX(nextConfig), {
  dest: 'public' // I don't know which plugin this belongs :/
})));
```

Having everything at the top:

```js
// GOOD EXAMPLE

const withBundleAnalyzer = require('@next/bundle-analyzer')({
    enabled: process.env.ANALYZE === 'true',
});
const withPWA = require('next-pwa')({
    dest: 'public'
});

module.exports = withBundleAnalyzer(withPWA(withMDX(nextConfig)));
```

**Comment by ymchun** [comment]
@AleksandarDev is the following also a good example?

```js
const withPWA = require('next-pwa')

let nextConfig = {
  // your next config goes here
}

nextConfig = withPWA(nextConfig, {
  // next-pwa config goes here
})

const withBundleAnalyzer = require('@next/bundle-analyzer')({
    enabled: process.env.ANALYZE === 'true',
});

const withMDX = require('@next/mdx')({
  options: {
    remarkPlugins: [],
    rehypePlugins: [],
  },
})

nextConfig = withBundleAnalyzer(withMDX(nextConfig));
// for the line above, I have a concern if there are many library to use, this line becomes a very long nested function call

module.exports = nextConfig
```

**Comment by AleksandarDev** [comment]
> @AleksandarDev is the following also a good example?

I personally don't like reassigning, that aside, why have 2 different ways of initializing plugin when we have official implementations as the reference?

**Comment by DavidSint** [comment]
Having implemented a fix in #368 to solve one problem, clearly I didn't expect these issues to arise. Perhaps some tests would be needed to ensure this could be detected earlier. Having looked into the next.js changes, I think it is [this](https://github.com/vercel/next.js/pull/38498) specific PR that brings in the validation that has warns against the way this  package has its configuration.

As part of the comments, @ijjk [says](https://github.com/vercel/next.js/pull/38498#issuecomment-1197282975):
> the pattern we recommend for plugins can be seen in our next-mdx package which doesn't require adding the plugin specific configs to the returned config itself

and they provide's the link: https://github.com/vercel/next.js/blob/e955850dd4880430433043fad3d8c13ed43b089a/packages/next-mdx/index.js#L1 

Based on these comments, I think the best thing this package can do to protect itself from future breaking changes is to copy the Next.js recommended way of writing a plugin API. If this package continues to roll its own way, Next.js could implement something else in future which would mean the function signature would need to change again.

So in summary, I agree with @AleksandarDev that the best approach is to copy the Next.js recommended plugin API signature.

**Comment by marcofranssen** [comment]
> Having implemented a fix in #368 to solve one problem, clearly I didn't expect these issues to arise. Perhaps some tests would be needed to ensure this could be detected earlier.
> 
> Having looked into the next.js changes, I think it is [this](https://github.com/vercel/next.js/pull/38498) specific PR that brings in the validation that has the warnings against the way this package has its configuration in the same place as the Next.js configuration.
> 
> As part of the comments of the PR, @ijjk [says](https://github.com/vercel/next.js/pull/38498#issuecomment-1197282975):
> 
> > the pattern we recommend for plugins can be seen in our next-mdx package which doesn't require adding the plugin specific configs to the returned config itself
> 
> and they provide the link: [vercel/next.js@`e955850`/packages/next-mdx/index.js#L1](https://github.com/vercel/next.js/blob/e955850dd4880430433043fad3d8c13ed43b089a/packages/next-mdx/index.js#L1)
> 
> Based on these comments, I don't see why this package should go against their recommendation. If this package continues to roll its own way, Next.js could implement something else in future which would mean the function signature would need to change again. Therefore, the best thing this package can do to protect itself from future breaking changes is to copy the Next.js recommended way of writing a plugin API.
> 
> So in summary, I agree with @AleksandarDev that the best approach is to copy the Next.js recommended plugin API signature.

Fully Agree.

However before introducing a breaking change I think it would be great if the plugin also has a fix for the current release https://github.com/shadowwalker/next-pwa/pull/384. Might make sense to merge this fix and then work on a refactor to follow the best practices.

**Comment by kirkegaard** [comment]
Its almost impossible to see what this fixes as the PR is filled with format changes. Can we please move the format changes to a separate PR, if those changes are really needed?
